source/common/init/target_impl.cc

```diff
@@ -27,19 +27,27 @@ TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
       fn_(std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
```

Comment:
This is the key spot I think you need to modify. I think it needs to look something like:

```c++
if (is_ready_) {
  // If we were already ready before being called by the init manager, just signal immediately...
  watcher_handle->ready();
} else {
  // ... Otherwise, hang onto the init manager's watcher handle and initialize.
  watcher_handle_ = std::move(watcher_handle);
  fn();
}
```

Reply:
Thanks! That is close to my plan B. Well, this is even better as I somehow considered `is_ready_` should be shared pointer. Will use your clean code :)

Reply:
@mergeconflict  It's arguable if the a ready target would invoke fn().

I prefer to keep the existing behavior that fn() would be invoked to . WDYT?

Reply:
I don't know, that seems a little sketchy to me, maybe... `fn()` is the target's initialization function, and if the target is already ready, wouldn't that mean `fn()` (or its equivalent) would have already been called before?

Reply:
For those user who call ready() only in fn(), both are fine.

For the users who may call ready() before fn(), the users will see fn() invoked prior to this PR. Since we don't have a strong opinion whether fn() should be called or not, let's not break the behavior that fn() is called in this situation.

For my use case, my fn in Target constructor is empty as the ready() depends on other signal. So I vote for NOT to assume ready() always comes from fn().

Reply:
Whoa... this doesn't make sense to me. If the initialization function for your target is empty, and your target is always ready immediately, it seems like you have no need for `Init::Manager` at all.

Reply:
Sorry for the unclear. 

> The initialization function is empty 

True

>  your target is always ready immediately

Not always. I don't want to introduce the complexity here so let me simplify the case I was seeing.
The listener could provide the server scope Init::Manager or per listener Init::Manager. The former one could be initialized really late but my target wants to be started as early as possible but it doesn't guarantee be ready before Init::Manager::initialize().

Reply:
Hmm, that's complicated... So, in this case, even though the init manager isn't telling the target _when_ to initialize, the init manager still needs to know that the target _has_ initialized, did I get that right?

Reply:
Exactly. Actually I think the existing CDS/LDS should follow this pattern as well: cluster should be a target of listener's init manager and cluster should be updated independently. However, CDS/LDS doesn't use this at the risk that throw the ordering to ADS or user.